tests/Test/Dscp/DB/SQLite/Common.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE ExplicitForAll             #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE PartialTypeSignatures      #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeSynonymInstances       #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Test.Dscp.DB.SQLite.Common
+       ( module Test.Common
+       , module Test.Dscp.DB.SQLite.Common
+       , module Dscp.Core.Types
+       , hash
+       ) where
+
+import Prelude hiding (fold)
+
+import Database.SQLite.Simple
+
+import System.Directory (removeFile)
+import System.IO.Error (IOError, isDoesNotExistError)
+
+import Test.QuickCheck.Gen (generate)
+
+import Dscp.Core.Types (
+    Address (..),
+    Assignment (..),
+    AssignmentType (..),
+    CourseId (..),
+    --Grade (..),
+    SignedSubmission (..),
+    --StudentId (..),
+    --SubjectId,
+    Submission (..),
+    SubmissionType (..),
+    SubmissionSig,
+    SubmissionWitness (..),
+    --aAssignment,
+    aCourseId,
+    --aType,
+    sAssignment,
+    sStudentId,
+    --sType,
+    ssSubmission,
+    ssWitness,
+    swKey
+    --swSig
+  )
+import qualified Dscp.DB.SQLite.Class as Adapter
+import Dscp.DB.SQLite.Schema (ensureSchemaIsSetUp)
+import Dscp.Crypto (
+    Hash,
+    PublicKey,
+    Signature,
+    HasSignature,
+    HasHash,
+    sign,
+    hash
+  )
+
+import Test.Common
+
+newtype TestSQLiteM a
+    = TestSQLiteM { getTestSQLiteM :: ReaderT Connection IO a }
+    deriving (Functor, Applicative, Monad, MonadIO, MonadThrow, MonadCatch)
+
+runTestSQLiteM :: TestSQLiteM a -> IO a
+runTestSQLiteM action = do
+    let filename = "test.db"
+
+    removeFile filename `catch` \(e :: IOError) -> do
+        if isDoesNotExistError e
+        then return ()
+        else throwM e
+
+    withConnection filename $ \conn -> do
+        ensureSchemaIsSetUp conn
+        setTrace conn (Just print)
+        getTestSQLiteM action `runReaderT` conn
+
+sqliteProperty
+    :: Testable prop
+    => Arbitrary a
+    => (a -> TestSQLiteM prop)
+    -> Property
+sqliteProperty action =
+    ioProperty $ do
+        input <- generate arbitrary
+        runTestSQLiteM (action input)
```

Comment:
If we don't remove it, we might end up generating the same input and some tests will break.

We might store test data in `":memory:"`, to speed things up.

Reply:
I made it to be on disk, because I wanted to, after my test fail, go into it and see what's wrong.

However `Database.SQLite.Simple.setTrace` and `sqlite-simple` exceptions made it quite simple to debug.

Reply:
Let's use `:memory:` storage indeed.